### PR TITLE
zsh autocomplete

### PR DESCRIPTION
**Pull request by jhheider**
This works. The places I think it might need love:
- it will be slow. and ever slower. useSync should cache all the provides to the sqlite store for much faster performance.
- ~it wasn't in any way clear to me how to interleave our completions with zsh's defaults. possibly because we don't work like anything else~
  - ~one possibility is to do the path scanning ourselves, so we can just do the whole completion stack. this will also take time, and want for caching.~

### CODE REVIEW

**Comment by jhheider** [comment]
Those errors are a little thorny. Looks like our zsh package might need:
```yaml
runtime:
  env:
    FPATH: ${{prefix}}/functions
    MODULE_PATH: ${{prefix}}/lib/zsh/{{version.marketing}}
```
or similar...

**Comment by jhheider** [comment]
I believe https://github.com/teaxyz/pantry/pull/2132 will fix this PR, if it's deemed and acceptable solution.

**Comment by mxcl** [comment]
why *is* this dependent on our zsh? zsh comes with mac. Yes, our zsh should work, and linux ofc. But we should be able to add this without fixing zsh and looking at my current schedule I will not be able to fix zsh for weeks.

**Comment by jhheider** [comment]
Probably because of how our tests work. I'll see if I can fix the test.

**Comment by jhheider** [comment]
https://github.com/teaxyz/pantry/pull/2132 should do it.

**Comment by jhheider** [comment]
Confirmed, and `zsh` fixed in the bargain. Ready for critique!

File: .vscode/settings.json [hunk 1]
```diff
@@ -5,5 +5,6 @@
   "deno.config": "deno.jsonc",
   "markdownlint.config": {
     "extends": "./.github/markdownlint.yml"
-  }
+  },
+  "editor.formatOnSaveMode": "modificationsIfAvailable"
```
Comment (mxcl) [inline .vscode/settings.json:9]: seems personal

File: src/app.magic.ts [hunk 2]
```diff
@@ -43,15 +43,37 @@ export default function(self: Path, shell?: string) {
...
         fi
       }
-      `
+
+      _tea_completion() {
+        local completions
+
+        if _has_tea_magic; then
+          # Call \`tea --complete\` with the current word as the prefix
+          completions=($(tea --complete \${words[CURRENT]}))
```
Comment (mxcl) [inline src/app.magic.ts:64]: even if we use sqlite having to use `tea` will add 80ms to all completion which sucks

File: src/app.magic.ts [hunk 3]
```diff
@@ -43,15 +43,37 @@ export default function(self: Path, shell?: string) {
...
+          compadd "$completions[@]"
+        fi
+
+        # Run the default completions no matter what
+        _complete
```
Comment (mxcl) [inline src/app.magic.ts:69]: does this mean we are also running for eg `git foo<TAB>`

**Review by mxcl — CHANGES_REQUESTED** [review]
Tried it and it works!

```
$ py
pygmentize        pyproject-build   python            
pylsp             pytest            python\{\{        
python                               python\{\{\ version.major\ \}\}
python3                              python\{\{\ version.marketing\ \}\}
```

needs some moustache expansion.

```
$ git cl
clang          clear          cluster        
clang++        clippy-driver  clusterdb      
clean -- remove untracked files from working tree
clone -- clone repository into new directory
```

Seems we are completing any time `TAB` is pressed.

**Comment by mxcl** [comment]
Can we have tea completions in teal?

File: src/app.magic.ts (hunk 3 above)
Comment (jhheider) [reply src/app.magic.ts:69]: I believe that's a different piece of issue. This is so things in $PATH are also listed appropriately.

File: .vscode/settings.json (hunk 1 above)
Comment (jhheider) [reply .vscode/settings.json:9]: can remove

**Comment by jhheider** [comment]
> Can we have tea completions in teal?

worth a try, certainly.

**Comment by jhheider** [comment]
> needs some moustache expansion.
> 

Hm. Might just need to omit mustachioed variants. Otherwise, we'd need to do produce them for every version available...

> Seems we are completing any time `TAB` is pressed.

That we should not do. Probably something I missed about detecting context.

**Comment by jhheider** [comment]
the docs for zsh completion are long but not amazing.

**Comment by jhheider** [comment]
colorizing the results eluded me. i believe everything else is addressed.

**Comment by mxcl** [comment]
> colorizing the results eluded me

was a nice to have, not required. good job trying tho

**Comment by mxcl** [comment]
So trying it out, being able to disntiguish between what is on the system and what is not yet on the system is pretty important IMO.

We either need coloration or to make the results appear completely separately to system results

**Comment by jhheider** [comment]
Zsh completion has a grouping system. I wasn't able to properly group output (complicated, like I said), but supposedly it can be done by tags, then print a header for each section.

I go back and forth about the import. If the point is to be magical, then not showing the difference seems like the right way. If they cared about local vs tea installed, they wouldn't use magic, right?

**Comment by mxcl** [comment]
> I go back and forth about the import. If the point is to be magical, then not showing the difference seems like the right way. If they cared about local vs tea installed, they wouldn't use magic, right?

it's a good point.

Let's just merge and see how it feels for a few weeks I guess.

**Review by mxcl — APPROVED** [review]

**Pull request merged**